### PR TITLE
Added support for Elasticsearch 6.x & 5.x

### PR DESCRIPTION
**Pull request by ael-code**
A lot of breaking changes have been introduced in ES version `5.x` and `6.x`. 
I manage to make the libreant code compatible with **ALL** the versions of Ealsticsearch with very very few if/else construct.

Travis still doesn't provide a straightforward way to install Elasticsearch `6.x` thus we are only testing till the `5.5` version.

### CODE REVIEW

**Comment by ael-code** [comment]
@leophys any news on this?

**Comment by ael-code** [comment]
Do you need any further information?

**Comment by leophys** [comment]
Sorry @ael-code 
I didn't get I was the assignee. I run the tests inside a docker container with es-6 running.
Am I missing something?

**EDIT:** Is this related to #330 ?

```
(ve) root@1402521e36e8:/libreant# python setup.py test
running test
running egg_info
writing requirements to libreant.egg-info/requires.txt
writing libreant.egg-info/PKG-INFO
writing top-level names to libreant.egg-info/top_level.txt
writing dependency_links to libreant.egg-info/dependency_links.txt
writing entry points to libreant.egg-info/entry_points.txt
reading manifest file 'libreant.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'libreant.egg-info/SOURCES.txt'
running build_ext
EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE.....EEEEEEEEEEE.....FF.FF................Traceback (most recent call last):
  File "setup.py", line 172, in <module>
    setup(**conf)
  File "/libreant/ve/local/lib/python2.7/site-packages/setuptools/__init__.py", line 129, in setup
    return distutils.core.setup(**attrs)
  File "/usr/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/libreant/ve/local/lib/python2.7/site-packages/setuptools/command/test.py", line 226, in run
    self.run_tests()
  File "setup.py", line 80, in run_tests
    nose.run_exit(argv=['nosetests'])
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/core.py", line 121, in __init__
    **extra_args)
  File "/usr/lib/python2.7/unittest/main.py", line 95, in __init__
    self.runTests()
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/core.py", line 207, in runTests
    result = self.testRunner.run(self.test)
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/core.py", line 62, in run
    test(result)
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/suite.py", line 177, in __call__
    return self.run(*arg, **kw)
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/suite.py", line 224, in run
    test(orig)
  File "/usr/lib/python2.7/unittest/suite.py", line 70, in __call__
    return self.run(*args, **kwds)
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/suite.py", line 75, in run
    test(result)
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/suite.py", line 177, in __call__
    return self.run(*arg, **kw)
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/suite.py", line 224, in run
    test(orig)
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/suite.py", line 177, in __call__
    return self.run(*arg, **kw)
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/suite.py", line 214, in run
    result.addError(self, self._exc_info())
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/proxy.py", line 128, in addError
    formatted = plugins.formatError(self.test, err)
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/plugins/manager.py", line 99, in __call__
    return self.call(*arg, **kw)
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/plugins/manager.py", line 141, in chain
    result = meth(*arg, **kw)
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/plugins/logcapture.py", line 237, in formatError
    return (ec, self.addCaptureToErr(ev, records), tb)
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/plugins/logcapture.py", line 244, in addCaptureToErr
    records + \
  File "/libreant/ve/local/lib/python2.7/site-packages/nose/util.py", line 651, in safe_str
    return str(val)
  File "/libreant/ve/local/lib/python2.7/site-packages/elasticsearch/exceptions.py", line 55, in __str__
    cause = ', %r' % self.info['error']['root_cause'][0]['reason']
TypeError: string indices must be integers
```

**Comment by ael-code** [comment]
@leophys wrote:
> Sorry @ael-code
> I didn't get I was the assignee. I run the tests inside a docker container with es-6 running.
> Am I missing something?

Your are probably using a version of the elasticsearch-py library that doesn't match the version of your running es cluster (you said to be `6`).

I've added an additional commit that performs this check at runtime, so now you should receive a proper exception that inform you about the version mismatch.

Can you try to run the new code before to install the correct version of elasticsearch-py? What do you get?

**Comment by leophys** [comment]
Thanks @ael-code 
I have to admit I am still not comfortable with nose (I only use pytest, my fault).
Digging into the `setup.py` I discovered the existance of the `ES_VERSION` env variable.
Doing
```bash
(ve) $ ES_VERSION=6 pip install -e .
```
and
```bash
(ve) $ ES_VERSION=6 python setup.py test
```
does the job (all tests passed).

I'd merge. Do you suggest some other investigation?

**Comment by ael-code** [comment]
@leophys wrote:
> I'd merge. Do you suggest some other investigation?

Can you confirm that the last commit works properly? It should provide an AssertionError when the two versions do not match.

I'd like also @boyska to give a review to the last commit. I'm not sure this is the best approach.

File: libreantdb/api.py
Comment (boyska) [inline libreantdb/api.py:364]: Doesn't this change require corresponding change in `archivant.insert_volume`? I mean, I know that the test passes, but I thought that changing a method prototype would have  required changes in callers.

File: utils/es.py
Comment (boyska) [inline utils/es.py]: I found it funny that a line that describes a _wrapper_ is not _wrapped_ to 72 columns :P

File: utils/es.py
Comment (boyska) [inline utils/es.py]: ## On subclassing

I would prefer not to do this. I think that a good solution might be the use of a "factory" function: that is, something that, using the exactly same logic of this code
* instantiate the official Elasticsearch class
* check if everything is right and either:
** raise an Exception (see below)
** returns the unmodified object

Subclassing should be used only when we really mean it.

## On AssertionError

btw, I don't think AssertionError is appropriate here; [doc](https://docs.python.org/3.5/library/exceptions.html#AssertionError) is clear that is reserved to the `assert` statement, which is not what we're doing.  As an example, `AssertionError`s would be surprising when running with `python -O`. I think that a `RuntimeError` can be good enough for our case

File: libreantdb/api.py
Comment (boyska) [inline libreantdb/api.py:17]: I hate them for changing this kind of trivialities.

**Review by boyska — CHANGES_REQUESTED** [review]
I like most of it, except the subclassing.

File: libreantdb/api.py
Comment (ael-code) [reply libreantdb/api.py:17]: How would you do that? Can you provide an alternative solution?

Comment (boyska) [reply libreantdb/api.py:17]: I hate _them_ (=ES devs) not you :) I think your solution is correct and clear

File: libreantdb/api.py
Comment (ael-code) [reply libreantdb/api.py:364]: I just specialized the function parameter according to the invocation inside `insert_volume`. Thus it is all coherent.

File: utils/es.py
Comment (ael-code) [reply utils/es.py]: :) Should be fixed now.

File: utils/es.py
Comment (ael-code) [reply utils/es.py]: I thought on this, the thing that I like of this approach is that it is very easy to revert. As you can see I needed to just change the import line in all the files that use the Elasticsearch class, but I didn't actually change any code.

About the AssertionError I didn't know that the python optimization skip those. I've change it to RuntimeError as you suggested.

**Comment by ael-code** [comment]
I think we are ready to merge. Waiting for your final approval.

File: utils/es.py
Comment (boyska) [reply utils/es.py]: > I thought on this, the thing that I like of this approach is that it is very easy to revert. As you can see I needed to just change the import line in all the files that use the Elasticsearch class, but I didn't actually change any code.

same applies if you change `class Elasticsearch` to `def Elasticsearch`. You can still keep code compatibility, returning an object that is the "true" one (in case of success).

> About the AssertionError I didn't know that the python optimization skip those.

That's not what I meant. Python optimization skips `asserts`; it doesn't ignore `AssertionError`s if those are raised by user code.
But what you're doing here is not an assert. Generally speaking, when an assert is violated it means that the programmer committed an error; that's why assert can be considered a debugging tool, or even a "documentation" tool, not an error-checking tool.
An assert should never check user input, system environment, or anything that is out of the control of the programmer.
What I meant is that raising an `AssertionError` would be confusing, because those are _unexpected_ with `python -O`. But this is just one of the consequences of using `AssertionError` for something that is not an `assert` :)

**Comment by boyska** [comment]
@ael-code I'm still itchy about that "class vs factory" thing. Read my reply; if you think my motivation doesn't hold, it's not a blocker.

**Comment by ael-code** [comment]
@boyska wrote:
> @ael-code I'm still itchy about that "class vs factory" thing. Read my reply; if you think my motivation doesn't hold, it's not a blocker.

I've already shared my opinion on that:

> I thought on this, I like this approach because is very easy to revert. As you can see I needed to just change the import line in all the files that use the Elasticsearch class, but I didn't actually change any code.

**Comment by boyska** [comment]
I thought I replied, but apparently I haven't. So let's see. Your point about not changing the code is good, but a factory would be as good in this. Let me give you an example

```python
from elasticsearch import Elasticsearch as Elasticsearch_official
def Elasticsearch(*args, **kwargs):
    check_version = kwargs.get('check_version', False)
    del kwargs['check_version']
    es = Elasticsearch(*args, **kwargs)
    if check_version:
        es_ver = es.info()['version']['number'].split('.')
        if(int(es_ver[0]) != int(es_pylib_version[0]):
            raise RuntimeError("Version mismatch")
    return es
```

This keeps compatibility without using unnecessary subclassing. Do you see any defect in this?

**Comment by ael-code** [comment]
@boyska wrote:
> This keeps compatibility without using unnecessary subclassing. Do you see any defect in this?

I didn't see any defect in either two of the methodologies. In any case with the last commit I've implemented the wrapper function ;)

**Comment by boyska** [comment]
now the only error is the use of past simple: I will always hate (unneeded) subclassing :D

**Pull request merged**
